domain/src/main/java/life/qbic/domain/events/DomainEventProducer.java

```diff
@@ -0,0 +1,60 @@
+package life.qbic.domain.events;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * <b><class short description - 1 Line!></b>
+ *
+ * <p><More detailed description - When to use, what it solves, etc.></p>
+ *
+ * @since <version tag>
+ */
+public class DomainEventProducer {
+
+  private static final ThreadLocal<List> subscribers = new ThreadLocal<>();
+
+  private static final ThreadLocal<Boolean> publishing = ThreadLocal.withInitial(
+      () -> Boolean.FALSE);
+
+  public static DomainEventProducer instance(){
+    return new DomainEventProducer();
+  }
+
+  public DomainEventProducer() {
+    super();
+  }
+
+  public <T extends DomainEvent> void subscribe(DomainEventSubscriber<T> subscriber) {
+    if (publishing.get()) {
+      return;
+    }
+    List<DomainEventSubscriber<T>> registeredSubscribers = subscribers.get();
+
+    if (registeredSubscribers == null) {
+      registeredSubscribers = new ArrayList<>();
+      subscribers.set(registeredSubscribers);
+    }
+
+    registeredSubscribers.add(subscriber);
+  }
+
+  public <T extends DomainEvent> void publish(final T domainEvent) {
+    if (publishing.get()) {
+      return;
+    }
+    try {
+      publishing.set(Boolean.TRUE);
+      List<DomainEventSubscriber<T>> registeredSubscribers = subscribers.get();
+      Class<? extends DomainEvent> domainEventType = domainEvent.getClass();
+      for (DomainEventSubscriber<T> subscriber : registeredSubscribers) {
+        if (subscriber.subscribedToEventType() == domainEventType) {
+          subscriber.handleEvent(domainEvent);
+        }
+      }
```

Comment:
NIT: personal flavor for readability
```suggestion
      registeredSubscribers.stream()
          .filter(it -> it.subscribedToEventType() == domainEventType)
          .forEach(DomainEventSubscriber::handleEvent):
```